sympy/parsing/mathematica.py

```diff
@@ -1,28 +1,101 @@
+import typing
+from itertools import product
+from typing import Any, Dict as tDict, Tuple as tTuple, List, Optional, Union as tUnion, Callable
+
+import sympy
+from sympy import Mul, Add, Pow, log, exp, sqrt, cos, sin, tan, asin, acos, acot, asec, acsc, sinh, cosh, tanh, asinh, \
+    acosh, atanh, acoth, asech, acsch, expand, im, flatten, polylog, cancel, expand_trig, sign, simplify, \
+    UnevaluatedExpr, S, atan, atan2, Mod, Max, Min, rf, Ei, Si, Ci, airyai, airyaiprime, airybi, primepi, prime, \
+    isprime, cot, sec, csc, csch, sech, coth, Function, I, pi, Tuple, GreaterThan, StrictGreaterThan, StrictLessThan, \
+    LessThan, Equality, Or, And, Lambda, Integer, Dummy, symbols
+from sympy.core.sympify import sympify, _sympify
+from sympy.functions.special.bessel import airybiprime
+from sympy.functions.special.error_functions import li
+from sympy.utilities.exceptions import SymPyDeprecationWarning
 
 
 def mathematica(s, additional_translations=None):
-    '''
-    Users can add their own translation dictionary.
-    variable-length argument needs '*' character.
+    """
+    Translate a string containing a Wolfram Mathematica expression to a SymPy
+    expression.
+
+    If the translator is unable to find a suitable SymPy expression, the
+    ``FullForm`` of the Mathematica expression will be output, using SymPy
+    ``Function`` objects as nodes of the syntax tree.
 
     Examples
     ========
 
     >>> from sympy.parsing.mathematica import mathematica
-    >>> mathematica('Log3[9]', {'Log3[x]':'log(x,3)'})
-    2
-    >>> mathematica('F[7,5,3]', {'F[*x]':'Max(*x)*Min(*x)'})
+    >>> mathematica("Sin[x]^2 Tan[y]")
+    sin(x)**2*tan(y)
+    >>> e = mathematica("F[7,5,3]")
+    >>> e
+    F(7, 5, 3)
+    >>> from sympy import Function, Max, Min
+    >>> e.replace(Function("F"), lambda *x: Max(*x)*Min(*x))
     21
 
-    '''
-
+    Both standard input form and Mathematica full form are supported:
+
+    >>> mathematica("x*(a + b)")
+    x*(a + b)
+    >>> mathematica("Times[x, Plus[a, b]]")
+    x*(a + b)
+
+    To get a matrix from Wolfram's code:
+
+    >>> m = mathematica("{{a, b}, {c, d}}")
+    >>> m
+    ((a, b), (c, d))
+    >>> from sympy import Matrix
+    >>> Matrix(m)
+    Matrix([
+    [a, b],
+    [c, d]])
+
+    If the translation into equivalent SymPy expressions fails, an SymPy
+    expression equivalent to Wolfram Mathematica's "FullForm" will be created:
+
+    >>> mathematica("x_.")
+    Optional(Pattern(x, Blank()))
+    >>> mathematica("Plus @@ {x, y, z}")
+    Apply(Plus, (x, y, z))
+    >>> mathematica("f[x_, 3] := x^3 /; x > 0")
+    SetDelayed(f(Pattern(x, Blank()), 3), Condition(x**3, x > 0))
+    """
     parser = MathematicaParser(additional_translations)
-    return sympify(parser.parse(s))
+
+    if additional_translations is not None:
+        SymPyDeprecationWarning(
+            feature="additional_translations parameter for the Mathematica parser",
+            last_supported_version="1.9",
```

Comment:
No idea. @oscarbenjamin ?

Reply:
Why are we removing additional_translations? This is a regression. Without this, it is impossible for people to define custom translations for their own functions. Is it not possible to keep support for this?

Reply:
> Without this, it is impossible for people to define custom translations for their own functions. Is it not possible to keep support for this?

New way to do it (much more simple and functional):
`mathematica(" ... ").subs(additional_translation)`

Reply:
If this is to be deprecated then it should say `last_supported_version="1.10"`.

Reply:
@oscarbenjamin isn't the release branch already made? If this is merged into master it will go into 1.11. At least release.py in `master` says 1.11.dev. In general the last_supported_version should be whatever is in release.py, minus `.dev`.

Reply:
Yes, so if this is merged then the feature will be deprecated in 1.11 and then the last supported version will be 1.10.

Reply:
Hmm, that's not how I've been interpreting this flag, and it's not how I documented it in https://github.com/sympy/sympy/pull/22900. I wonder if others have been doing both. I guess the flag should be renamed to `deprecated_since_version` to make this clearer.

Reply:
Oh I see. There is `deprecated_since_version` *and* `last_supported_version`. So yeah, this is something I already fixed in https://github.com/sympy/sympy/pull/22900. I removed `last_supported_version` and there's now just `deprecated_since_version`. `last_supported_version` is confusing and completely redundant, and exactly 0 deprecations in master are using it.

So in short, replace this with `deprecated_since_version="1.11"`.

If you want to wait until https://github.com/sympy/sympy/pull/22900 is merged and do this deprecation in the new style, that would be great too. Otherwise if this is merged first, I'll update it in that PR.

Reply:
By the way, "last_supported_version" was actually supposed to be the last version that would support the feature, i.e., the future version. But we don't even use a version-based deprecation policy, which is why I removed it.